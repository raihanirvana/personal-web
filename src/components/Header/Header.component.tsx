import React from 'react';

import {
  FaHome,
  FaInfoCircle,
  FaBriefcase,
  FaProjectDiagram,
} from "react-icons/fa";

import { HeaderList } from "./Header.type";

const headers: HeaderList = [
  { key: 1, title: "Home", icon: <FaHome /> },
  { key: 2, title: "About", icon: <FaInfoCircle /> },
  { key: 3, title: "Resume", icon: <FaBriefcase /> },
  { key: 4, title: "Project", icon: <FaProjectDiagram /> },
];

/**
 * Renders a header element with optional styles for first or last headers.
 * 
 * @param {string} title - The title text for the header.
 * @param {JSX.Element} icon - The title text for the header.
 * @returns {JSX.Element} A JSX element representing the header with the specified title and styling.
 */
const renderHeader = (
  title: string,
  icon: JSX.Element,
): JSX.Element => (
  <div
    className={`w-10 text-white overflow-hidden text-center p-4 pl-2 hover:overflow-visible hover:w-26 transition-all duration-300 ease-in-out hover:cursor-pointer`}
  >
    <button className="flex items-center justify-center gap-4 hover:cursor-pointer">
      <span className="text-xl mb-1">{icon}</span>
      <span className="text-sm">{title}</span>
    </button>
  </div>
);
/**
 * A functional component that renders a list of headers.
 * 
 * @returns {JSX.Element} A JSX element containing the header list.
 */
const Header = (): JSX.Element => (
  <div className="fixed top-1/2 left-0 transform -translate-y-1/2 ml-10">
    {headers.map(({ title, icon, key }) => (
      <React.Fragment key={key}>
        {renderHeader(title, icon)}
      </React.Fragment>
    ))}
  </div>
);

export default Header;