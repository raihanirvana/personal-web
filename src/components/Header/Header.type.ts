export type HeaderItem = {
  key: number;
  icon: JSX.Element;
  title: string;
};

export type HeaderList = HeaderItem[];

