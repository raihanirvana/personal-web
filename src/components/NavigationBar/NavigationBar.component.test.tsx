import { render, screen, fireEvent, renderHook, act } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import Header, { useNavbarVisibility } from './NavigationBar.component';

jest.useFakeTimers();

describe('Header Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Header activeBar={1} />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should apply opacity-20 while scrolling and revert to opacity-100 after debounce', () => {
    const { container } = render(
      <MemoryRouter>
        <Header activeBar={1} />
      </MemoryRouter>
    );
  
    const headerWrapper = container.querySelector('div.flex');
    expect(headerWrapper?.className).toContain('opacity-100');
    expect(headerWrapper?.className).not.toContain('opacity-20');
  
    act(() => {
      window.dispatchEvent(new Event('scroll'));
    });
  
    expect(headerWrapper?.className).toContain('opacity-20');
    expect(headerWrapper?.className).not.toContain('opacity-100');
  
    act(() => {
      jest.advanceTimersByTime(500);
    });
  
    expect(headerWrapper?.className).toContain('opacity-100');
    expect(headerWrapper?.className).not.toContain('opacity-20');
  });

  it('navigates to the correct route when a link is clicked', async () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path="/" element={<Header activeBar={1} />} />
          <Route path="/resume" element={<div>Resume</div>} />
        </Routes>
      </MemoryRouter>
    );

    const aboutLink = screen.getByRole('link', { name: /resume/i });
    fireEvent.click(aboutLink);

    expect(await screen.findByText('Resume')).toBeInTheDocument();
  });

  describe('useNavbarVisibility', () => {
    beforeEach(() => {
      jest.clearAllTimers();
    });
  
    it('should be false by default', () => {
      const { result } = renderHook(() => useNavbarVisibility());
      expect(result.current).toBe(false);
    });
  
    it('should become true on scroll and revert to false after delay', () => {
      const { result } = renderHook(() => useNavbarVisibility(1000));
  
      act(() => {
        window.dispatchEvent(new Event('scroll'));
      });
  
      expect(result.current).toBe(true);
  
      act(() => {
        jest.advanceTimersByTime(999);
      });
      expect(result.current).toBe(true);
  
      act(() => {
        jest.advanceTimersByTime(1);
      });

      expect(result.current).toBe(false);
    });
  
    it('should reset debounce timer if scroll happens again within delay', () => {
      const { result } = renderHook(() => useNavbarVisibility(1000));
  
      act(() => {
        window.dispatchEvent(new Event('scroll'));
      });

      expect(result.current).toBe(true);
  
      act(() => {
        jest.advanceTimersByTime(500);
        window.dispatchEvent(new Event('scroll'));
      });
  
      act(() => {
        jest.advanceTimersByTime(999);
      });

      expect(result.current).toBe(true);
  
      act(() => {
        jest.advanceTimersByTime(1);
      });

      expect(result.current).toBe(false);
    });
  });
});
