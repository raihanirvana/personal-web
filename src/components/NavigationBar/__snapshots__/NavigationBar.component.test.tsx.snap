// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="fixed top-1/2 left-5 transform -translate-y-1/2"
  >
    <button
      class="w-10 text-white overflow-hidden text-center p-4 flex gap-4 items-center pb-6 overflow-visible w-36 transition-all duration-300 ease-in-out cursor-pointer pl-8"
      type="button"
    >
      <span
        class="text-2xl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          />
        </svg>
      </span>
      <span
        class="text-xl"
      >
        Home
      </span>
    </button>
    <button
      class="w-10 text-white overflow-hidden text-center p-4 flex gap-4 items-center pb-6 hover:overflow-visible hover:w-36 transition-all duration-300 ease-in-out hover:cursor-pointer hover:pl-8"
      type="button"
    >
      <span
        class="text-2xl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
          />
        </svg>
      </span>
      <span
        class="text-xl"
      >
        About
      </span>
    </button>
    <button
      class="w-10 text-white overflow-hidden text-center p-4 flex gap-4 items-center pb-6 hover:overflow-visible hover:w-36 transition-all duration-300 ease-in-out hover:cursor-pointer hover:pl-8"
      type="button"
    >
      <span
        class="text-2xl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 336c0 8.84-7.16 16-16 16h-96c-8.84 0-16-7.16-16-16v-48H0v144c0 25.6 22.4 48 48 48h416c25.6 0 48-22.4 48-48V288H320v48zm144-208h-80V80c0-25.6-22.4-48-48-48H176c-25.6 0-48 22.4-48 48v48H48c-25.6 0-48 22.4-48 48v80h512v-80c0-25.6-22.4-48-48-48zm-144 0H192V96h128v32z"
          />
        </svg>
      </span>
      <span
        class="text-xl"
      >
        Resume
      </span>
    </button>
    <button
      class="w-10 text-white overflow-hidden text-center p-4 flex gap-4 items-center pb-6 hover:overflow-visible hover:w-36 transition-all duration-300 ease-in-out hover:cursor-pointer hover:pl-8"
      type="button"
    >
      <span
        class="text-2xl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M384 320H256c-17.67 0-32 14.33-32 32v128c0 17.67 14.33 32 32 32h128c17.67 0 32-14.33 32-32V352c0-17.67-14.33-32-32-32zM192 32c0-17.67-14.33-32-32-32H32C14.33 0 0 14.33 0 32v128c0 17.67 14.33 32 32 32h95.72l73.16 128.04C211.98 300.98 232.4 288 256 288h.28L192 175.51V128h224V64H192V32zM608 0H480c-17.67 0-32 14.33-32 32v128c0 17.67 14.33 32 32 32h128c17.67 0 32-14.33 32-32V32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </span>
      <span
        class="text-xl"
      >
        Project
      </span>
    </button>
  </div>
</DocumentFragment>
`;
